packages/stack/blockchain/src/index.js

```diff
@@ -18,6 +18,8 @@ export default class Blockchain {
     this.startedClient = null;
     this.plugins = options.plugins;
     this.blockchainClients = {};
+    this.warnIfPackageNotDefinedLocally = options.warnIfPackageNotDefinedLocally ?? warnIfPackageNotDefinedLocally;
+    this.Web3 = options.Web3 ?? Web3;
```

Comment:
You can't stub the way it was before:
```
import "Web3" from "web3";
class Blah {
  constructor(options) {

  }
}
```
The way this is, there is no web3 method available to stub.

Without some form of formal dependency injection, this is the only way that it can be done that I'm aware of. Do you have any other possible suggestions?

Reply:
Stubbing the constructor would be useful for unit testing the constructor **of our object under test**, without actually running the constructor function.

Reply:
Ok, I tried to do it and I can't even import for some reason. Whatever, it's all good

Reply:
Not sure if this is relevant to @emizzle's comment re: the constructor, but just in case...

If you have a class `Foo` and you want to create an instance of it without actually running the constructor, you can do:

```js
const foo = Object.create(Foo.prototype);
```

That is, you don't necessarily need to stub/mock the constructor.

Now assume that you want to test `meth1` method in isolation but `meth1` calls `meth2` and it expects that properties `.bar` and `.baz`, which would normally be set by the constructor, will be there. You can do something like this in a `beforeAll` or `beforeEach` within `describe('Foo#meth1', () => {...})`:

```js
foo.meth2 = jest.fn(() => {});
foo.bar = ...; // whatever value makes sense for the test
foo.baz = ...; // whatever value makes sense for the test
```

Note also that jest has built-in very powerful mocking capabilities, rivaling sinon.

See:
* https://jestjs.io/docs/en/mock-functions
* https://jestjs.io/docs/en/manual-mocks
* https://jestjs.io/docs/en/es6-class-mocks

There are other mocking-related pages in the Guides section as well.

Reply:
Not sure if this is relevant to @emizzle's comment re: the constructor, but just in case...

If you have a class `Foo` and you want to create an instance of it without actually running the constructor, you can do:

```js
const foo = Object.create(Foo.prototype);
```

That is, you don't necessarily need to stub/mock the constructor.

Now assume that you want to test `meth1` method in isolation but `meth1` calls `meth2` and it expects that properties `.bar` and `.baz`, which would normally be set by the constructor, will be there. You can do something like this in a `beforeAll` or `beforeEach` within `describe('Foo#meth1', () => {...})`:

```js
foo.meth2 = jest.fn(() => {});
foo.bar = ...; // whatever value makes sense for the test
foo.baz = ...; // whatever value makes sense for the test
```

Note also that jest has built-in very powerful mocking capabilities, rivaling sinon.

See:
* https://jestjs.io/docs/en/mock-functions
* https://jestjs.io/docs/en/manual-mocks
* https://jestjs.io/docs/en/es6-class-mocks

There are other mocking-related pages in the Guides section as well.

Reply:
Thanks for all the detail @michaelsbradleyjr. Definitely something to keep in mind for future tests! 

The issue here isn't running code in the constructor. It is more to do with stubbing an imported constructor `Web3`. Without injecting it in the to the class, it is difficult to test.